cdc/sinkv2/eventsink/mq/producer/kafka_producer.go

```diff
@@ -0,0 +1,243 @@
+		Value:     sarama.ByteEncoder(message.Value),
+		Metadata:  messageMetaData{callback: message.Callback},
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case k.asyncProducer.Input() <- msg:
+	}
+	return nil
+}
+
+func (k *kafkaProducer) Close() {
+	// We have to hold the lock to prevent write to closed producer.
+	k.closedMu.Lock()
+	defer k.closedMu.Unlock()
+	// If the producer was already closed, we should skip the close operation.
+	if k.closed {
+		// We need to guard against double closed the clients,
+		// which could lead to panic.
+		log.Warn("kafka producer already closed",
+			zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID),
+			zap.Any("role", k.role))
+		return
+	}
+	close(k.failpointCh)
+	// Notify the run loop to exit.
+	close(k.closedChan)
+	k.closed = true
+	// We need to close it asynchronously.
+	// Otherwise, we might get stuck with it in an unhealthy state of kafka.
+	go func() {
+		// `client` is mainly used by `asyncProducer` to fetch metadata and other related
+		// operations. When we close the `kafkaSaramaProducer`, TiCDC no need to make sure
+		// that buffered messages flushed.
+		// Consider the situation that the broker does not respond, If the client is not
+		// closed, `asyncProducer.Close()` would waste a mount of time to try flush all messages.
+		// To prevent the scenario mentioned above, close client first.
+		start := time.Now()
+		if err := k.client.Close(); err != nil {
+			log.Error("close sarama client with error", zap.Error(err),
+				zap.Duration("duration", time.Since(start)),
+				zap.String("namespace", k.id.Namespace),
+				zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+		} else {
+			log.Info("sarama client closed", zap.Duration("duration", time.Since(start)),
+				zap.String("namespace", k.id.Namespace),
+				zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+		}
+
+		start = time.Now()
+		err := k.asyncProducer.Close()
+		if err != nil {
+			log.Error("close async client with error", zap.Error(err),
+				zap.Duration("duration", time.Since(start)),
+				zap.String("namespace", k.id.Namespace),
+				zap.String("changefeed", k.id.ID),
+				zap.Any("role", k.role))
+		} else {
+			log.Info("async client closed", zap.Duration("duration", time.Since(start)),
+				zap.String("namespace", k.id.Namespace),
+				zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+		}
+	}()
+}
+
+func (k *kafkaProducer) run(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-k.closedChan:
+			return nil
+		case err := <-k.failpointCh:
+			log.Warn("receive from failpoint chan", zap.Error(err),
+				zap.String("namespace", k.id.Namespace),
+				zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+			return err
+		case ack := <-k.asyncProducer.Successes():
+			if ack != nil {
+				callback := ack.Metadata.(messageMetaData).callback
+				if callback != nil {
+					callback()
```

Comment:
What's the guarantee for the `callback()`? Some callbacks will never be called. Will this be a problem?

Reply:
If there is never an ack meaning the send fails, we first retry, and if the retry also fails, an error is reported eventually causing the processor to shut down and restart, so it is safe.

Reply:
>  we first retry

Who initiates the retry?

Reply:
Sarama producer.